app/models/pacbio/library.rb

```diff
@@ -93,6 +93,23 @@ def sequencing_plates
       wells&.collect(&:plate)
     end
 
+    # Return the data to be published as the source of an aliquot
+    def publish_data_source
+      {
+        sourceType: 'library',
+        sourceBarcode: tube&.barcode,
+        sampleName: request.sample_name
+      }
```

Comment:
My concern with doing this here is that you're tying the Traction Service models to one particular Avro message schema.  We made this mistake in the TOL-lab-share system when it pushes messages out to Traction.  Ideally I think you should have a separate class for turning your model into the specific Avro schema format.

If that could be something config driven, a schema for how to map values, that would be ideal, but might be a bit much for a first attempt at this.

Reply:
I think it might be worth going through this to see what we can do.

Reply:
@sdjmchattie  @stevieing I can see the point where Avro file format has significantly influenced  model objects.
I started with an approach similar to yours, where  a class to generates a message that conforms to the Avro format, but not using config. After introducing config, it was handed over the responsibility to model objects as which was the main purpose or effect of using config. 
So, you are suggesting  merging these two ideas?  Specifically, using a config driven approach where config delegates to a specific class responsible for generating this message. Is tat correct?

Reply:
I worked on a similar mapping problem about 7 years ago where it was in C# and we used something called AutoMapper.  I don't remember the specific, but I have some recollection of having a class that deals exclusively with one message type you want to create, and you were able to define what it should do via a config something a bit like:

```
library.tube&.barcode => sourceBarcode
library.request.sample_name => sampleName
```

I'm not sure whether there's any gem we can use in Ruby to support such a mapping operation.  What I've shown above is very much an oversimplification as I know there are arrays and other transformations that values might need to have done to them, so I don't think we should roll our own.  However, even if you just had a message generator class that you pass it the library, aliquot, whatever else is needed to fully form the message.  It would be OK for that class to know how to dissect the model and get the data needed.

It would be worthwhile commenting in the models that this class exists so any changes should be cross referenced to make sure the class is able to continue doing its job.  That would probably be caught by good unit tests though.

Another considering could be to have extensions for this class that make it mappable to the desired Avro format.  That way you keep the main model file clean but `include` the extensions to gain access to internal data.  Not sure if this is just adding more complexity though.

Reply:
Ok, There is already a class responsible for the mapping operation that we created. It may be best to extend this class so that it can handle the specific job of delegating from where these model methods are accessed. This approach would likely be the most viable starting point, I think.

Reply:
The refactoring includes a config-based approach to create the message without modifying model objects. This approach uses the yaml configuration to dynamically create an object responsible for message building, using a class specified in the configuration. It delegates the value creation of fields specified in the config to this object.  
The Avro schema version is included in the configuration to ensure that the schema it is created against is the same one configured in bunny. 
I hope that anything in the future that uses the 'EMQ::Publisher' can follow a similar structure, with new sections added against schema keys and versions in yml file.